File: api-server/src/domain/availabilityAggregator.ts
// Implemented for spec: agent/specs/meal-appointment-view-appointment-backend-implementation-spec.md

export interface AvailabilitySelection {
  participantId: string;
  slotKey: string;
}

export interface AvailabilityAggregationResult {
  availableCountBySlotKey: Map<string, number>;
  totalSelections: number;
}

export class AvailabilityAggregator {
  aggregate(selections: AvailabilitySelection[]): AvailabilityAggregationResult {
    const availableCountBySlotKey = new Map<string, number>();
    for (const selection of selections) {
      const current = availableCountBySlotKey.get(selection.slotKey) ?? 0;
      availableCountBySlotKey.set(selection.slotKey, current + 1);
    }

    return {
      availableCountBySlotKey,
      totalSelections: selections.length
    };
  }
}

File: api-server/internal/domain/availability_aggregator.go
// Implemented for spec: agent/specs/meal-appointment-view-appointment-backend-implementation-spec.md

package domain

type AvailabilitySelection struct {
	ParticipantID string
	SlotKey       string
}

type AvailabilityAggregationResult struct {
	AvailableCountBySlotKey map[string]int
	TotalSelections         int
}

type AvailabilityAggregator struct{}

func NewAvailabilityAggregator() *AvailabilityAggregator { return &AvailabilityAggregator{} }

func (AvailabilityAggregator) Aggregate(selections []AvailabilitySelection) AvailabilityAggregationResult {
	counts := make(map[string]int, len(selections))
	for _, s := range selections {
		counts[s.SlotKey]++
	}
	return AvailabilityAggregationResult{
		AvailableCountBySlotKey: counts,
		TotalSelections:         len(selections),
	}
}
